commands/http/handler.go

```diff
@@ -106,20 +105,18 @@ func (i internalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-	ctx, cancel := context.WithCancel(node.Context())
-	defer cancel()
-	/*
-		TODO(cryptix): the next line looks very fishy to me..
-		It looks like the the context for the command request beeing prepared here is shared across all incoming requests..
 
-		I assume it really isn't because ServeHTTP() doesn't take a pointer receiver, but it's really subtule..
-
-		Shouldn't the context be just put on the command request?
+	ctx, err := cmds.GetContext(node.Context(), req)
+	if err != nil {
+		err = fmt.Errorf("error parsing timeout option: %s", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 
-		ps: take note of the name clash - commands.Context != context.Context
-	*/
-	i.ctx.Context = ctx
-	req.SetContext(i.ctx)
+	//ps: take note of the name clash - commands.Context != context.Context
+	cmdctx := i.ctx
+	cmdctx.Context = ctx
+	req.SetContext(cmdctx)
```

Comment:
this whole operation is kinda funky. it's a lot of shufflling and the interface isnt clear. 

since we're pulling the context out of the request, cant we do this elsewhere, more globally? i.e. instead of saying:

```
req.Context().Context
```
we could rename `req.Context()` to `req.InvocationCtx()`  and use

```
req.Context()
```
to get a context derived from the node, with the timeout already applied.